go/fixchain/logger_test.go

```diff
@@ -0,0 +1,208 @@
+		// Check logger caching.
+		if test.chain != nil {
+			if test.ferr.Type == None && !l.postCertCache.get(hash(GetTestCertificateFromPEM(t, test.chain[0]))) {
+				t.Errorf("#%d: leaf certificate not cached", i)
+			}
+			if !l.postChainCache.get(hashChain(extractTestChain(t, i, test.chain))) {
+				t.Errorf("#%d: chain not cached", i)
+			}
+		}
+	}
+}
+
+// NewLogger() test
+func TestNewLoggerCaching(t *testing.T) {
+	// Test logging multiple chains by looking at caching.
+	newLoggerTest := struct {
+		url          string
+		chains       [][]string
+		expectedErrs []errorType
+	}{
+		"https://ct.googleapis.com/pilot",
+		[][]string{
+			{googleLeaf, thawteIntermediate, verisignRoot},
+			{googleLeaf, thawteIntermediate, verisignRoot},
+			{googleLeaf, thawteIntermediate},
+			{testLeaf, testIntermediate2, testIntermediate1, testRoot},
+		},
+		[]errorType{},
+	}
+
+	errors := make(chan *FixError)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go testErrors(t, 0, newLoggerTest.expectedErrs, errors, &wg)
+
+	l := NewLogger(5, newLoggerTest.url, errors, &http.Client{Transport: &newLoggerTestRoundTripper{}}, ratelimiter.NewLimiter(1000), false)
+
+	for _, chain := range newLoggerTest.chains {
+		l.QueueChain(extractTestChain(t, 0, chain))
+	}
+	l.Wait()
+	close(l.errors)
+	wg.Wait()
+
+	// Check logger caching.
+	seen := make(map[[hashSize]byte]bool)
+	for i, chain := range newLoggerTest.chains {
+		leafHash := hash(GetTestCertificateFromPEM(t, chain[0]))
+		if !l.postCertCache.get(leafHash) {
+			t.Errorf("Chain %d: leaf certificate not cached", i)
+		}
+		if !seen[leafHash] && !l.postChainCache.get(hashChain(extractTestChain(t, 0, chain))) {
+			t.Errorf("Chain %d: chain not cached", i)
+		}
+		seen[leafHash] = true
+	}
+}
+
+// Logger.postServer() test
+func TestPostServer(t *testing.T) {
+	for i, test := range postTests {
+		errors := make(chan *FixError)
+		l := &Logger{
+			url:            test.url,
+			client:         &http.Client{Transport: &postTestRoundTripper{t: t, test: &test, testIndex: i}},
+			toPost:         make(chan *toPost),
+			errors:         errors,
+			limiter:        ratelimiter.NewLimiter(1000),
+			postCertCache:  newLockedMap(),
+			postChainCache: newLockedMap(),
+		}
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go testErrors(t, i, test.expectedErrs, errors, &wg)
+
+		go l.postServer()
+		l.QueueChain(extractTestChain(t, i, test.chain))
+		l.Wait()
+
+		close(l.errors)
+		wg.Wait()
+	}
+}
+
+// Logger.IsPosted() test
+func TestIsPosted(t *testing.T) {
+	isPostedTests := []struct {
+		cert     string
+		expected bool
+	}{
+		{
+			googleLeaf,
+			true,
+		},
+		{
+			megaLeaf,
+			true,
+		},
+		{
+			testLeaf,
+			false,
+		},
+		{
+			testC,
+			false,
+		},
+	}
+
+	l := &Logger{postCertCache: newLockedMap()}
+	l.postCertCache.set(hash(GetTestCertificateFromPEM(t, googleLeaf)), true)
+	l.postCertCache.set(hash(GetTestCertificateFromPEM(t, megaLeaf)), true)
+	l.postCertCache.set(hash(GetTestCertificateFromPEM(t, testLeaf)), false)
```

Comment:
We don't in our code no, but that is how one would 'remove' an item from the postCertCache if they were going to.  So I was testing that IsPosted(item) would return false in the case that either the item is not in the cache at all, or it has been 'removed' by being set to false.